### PR TITLE
WIP: Add GPU aware MPI support in cannon algorithm

### PR DESCRIPTION
**Pull request by gsitaram**
If CUDA or HIP backend is enabled, `--WITH-DBCSR-G2G` CMake option enables the following:
- Use device arrays in MPI transfers in the cannon algorithm implementation.
- Move norms calculation to GPU
- Transfer the matrix images once to GPU and transpose the B matrix once. All MPI Isend and Irecv calls will transfer the data in device buffers instead of host buffers.
- MPI transfers and multiplications remain overlapped

Requirements:
- An MPI implementation that supports GPU aware communication.

Need help with the following:
- The OpenCL build fails with "undefined reference to `c_calculate_norms`" error, @hfp, could you help me fix it for the OpenCL backend?
- The ROCm build fails with undefined references in OpenMPI. I am guessing we have to fix the CI build to include ROCm aware support when building UCX that OpenMPI depends on.
- I need to include AMD copyright lines in each file modified, but this does not seem acceptable for the pre-commit script. Please let me know how to get around this.

### CODE REVIEW

**Comment by hfp** [comment]
> The OpenCL build fails with "undefined reference to c_calculate_norms" error, @hfp, could you help me fix it for the OpenCL backend?

Yes, I will help with this. I think it should not be a blocker for merge.

---

Btw, is there a way to perform atomic FP-add on (certain) AMD GPUs using OpenCL? It seems the CUDA based code path does that but everything I tried in OpenCL did not work for me like C11 atomics, legacy builtins, or even cheating by guessing a prototype function and calling it. I found some doc for inline assembly in OpenCL but I am a bit hesitant to adopt it (like reading all the arch specific doc for AMD GPUs). For NVidia btw I am using PTX inline assembly in OpenCL.

**Comment by hfp** [comment]
Regarding the failing check of file headers, I propose to put copyright lines in front of the LICENSE file like [this](https://github.com/libxsmm/libxsmm/blob/main/LICENSE.md), and to stick with generic

```text
! Copyright (C) by the DBCSR developers group - All rights reserved                                !
```

Otherwise we are back to the business of duplicating author notes as part of the source code. I think authorship is recorded as part of repository's metadata and no one is "left behind". As an extension, I also propose to drop AUTHORS file.

A general overhaul is to have a file extension `.md` for LICENSE (and perhaps similar files).
   
( sorry the proposal is not exactly related and can be perhaps a separate PR )

**Comment by gsitaram** [comment]
> > The OpenCL build fails with "undefined reference to c_calculate_norms" error, @hfp, could you help me fix it for the OpenCL backend?
> 
> Yes, I will help with this. I think it should not be a blocker for merge.
Thank you so much.
> 
> Btw, is there a way to perform atomic FP-add on (certain) AMD GPUs using OpenCL? It seems the CUDA based code path does that but everything I tried in OpenCL did not work for me like C11 atomics, legacy builtins, or even cheating by guessing a prototype function and calling it. I found some doc for inline assembly in OpenCL but I am a bit hesitant to adopt it (like reading all the arch specific doc for AMD GPUs). For NVidia btw I am using PTX inline assembly in OpenCL
I will find out more details and inform you.

**Comment by gsitaram** [comment]
> Regarding the failing check of file headers, I propose to put copyright lines in front of the LICENSE file like [this](https://github.com/libxsmm/libxsmm/blob/main/LICENSE.md), and to stick with generic
> 
> ```
> ! Copyright (C) by the DBCSR developers group - All rights reserved                                !
> ```
> 
> Otherwise we are back to the business of duplicating author notes as part of the source code. I think authorship is recorded as part of repository's metadata and no one is "left behind". As an extension, I also propose to drop AUTHORS file.
> 
> A general overhaul is to have a file extension `.md` for LICENSE (and perhaps similar files).
> 
> ( sorry the proposal is not exactly related and can be perhaps a separate PR )

I like this proposal of having a separate LICENSE file. If all DBCSR maintainers agree, I can make the required changes (keep only the DBCSR Developers Group copyright line in each file, and add a new license file with the entire text of the license and copyright lines above that.)

**Comment by gsitaram** [comment]
@hfp, it does not seem like there are atomic add operations in OpenCL for FP type data. You may have to use compare and swap instead to accomplish the same. I found [this blog](https://streamhpc.com/blog/2016-02-09/atomic-operations-for-floats-in-opencl-improved/) showing an example, but this is not recent. 
I see that in the newer OpenCL (3.0) spec, the atomic_* operations are deprecated, and we have to use a combination of memory order and fence operations.. but I am not very familiar with this set of functions. If I find any recent examples somewhere, I'll be sure to point those to you.

**Comment by hfp** [comment]
Please allow me to share my suggestions for this PR:

* Let's decide on the [license banner etc](https://github.com/cp2k/dbcsr/pull/647#issuecomment-1375812743) solely based on input from @alazzaro. I would avoid surveying all/past authors. Also, the suggested change is not about dropping anyone's contribution; it's fully recorded in the GitHub repository and we only avoid duplicating this information along with exposing potentially outdated contact data.
* I adjusted upstream master to provide a stub implementation of `c_calculate_norms` for the OpenCL backend. I hope I got the return code correct that allows to bail-out at runtime and to hopefully fall-back to host as long as c_calculate_norms is not implemented with an own OpenCL kernel. So, please consider to rebase or merge the upstream master.
* I suggest rethinking the file extension of `calculate_norms.cpp`. I believe `.cu` might require less changes in the build system(s). However, this is not too important.


In my integration test (#649), I changed including some header files in `calculate_norms.cpp` from:

```cpp
#if defined(__CUDA)
#  include "acc/cuda/acc_cuda.h"
#elif defined(__HIP)
#  include "acc/hip/acc_hip.h"
#endif
#include "libsmm_acc_init.h"
```

... to the following:

```cpp
#if defined(__CUDA)
#  include "../cuda/acc_cuda.h"
#elif defined(__HIP)
#  include "../hip/acc_hip.h"
#endif
#include "libsmm_acc_init.h"
```

**Comment by gsitaram** [comment]
@hfp, I merged the master into this branch and changed the path to those header files as you suggested. 
As for the name of the file, hipcc takes .cpp extension, and nvcc can compile the .cpp file with the option `-x cuda`. So, I didn't change it. If the changes introduced to the build system are not acceptable, please feel free to change. After I got it to build, I didn't touch it.
I'll discuss with Alfio and see how to proceed with the license blob.
Thanks for your help in getting the opencl backend to accept these changes.

**Comment by alazzaro** [comment]
@gsitaram Thanks for this PR and sorry for the late reply (I would like to thank @hfp for his contribution too).

I'm putting here some of the topics/comments:

- [ ] Definitely I agree with @hfp proposal, so we can change our LICENSE file. @dev-zero comments?
- [ ] Understand how we can test it. Currently we cannot on github, so we can use the CI on Daint
- [ ] Move file calculate_norms.cpp to cuda_hip directory

I will add some specific comments in the files.

File: src/mm/dbcsr_mm_common.F [hunk 1]
```diff
@@ -93,13 +104,32 @@ MODULE dbcsr_mm_common
...
    PUBLIC :: acc_transpose_blocks
+   PUBLIC :: acc_calculate_norms
 
    PUBLIC :: product_matrix_size_guess
 
    PUBLIC :: calculate_norms
    PUBLIC :: huge_norm
    PUBLIC :: local_filter
 
+#if defined (__DBCSR_ACC)
```
[the remark this answers is not in the document]
Comment (gsitaram) [reply src/mm/dbcsr_mm_common.F:115]: Done.

File: src/acc/libsmm_acc/libsmm_acc.cpp [hunk 2]
```diff
@@ -140,8 +141,8 @@ inline void jit_kernel(ACC_DRV(function) & kern_func, libsmm_acc_algo algo, int
   const char* compileOptions[] = {"-D__CUDA", "-w", ARCH_OPTION};
   size_t nOptions = 3;
 #else
-  const char* compileOptions[] = {"-D__HIP", "-O3", "-w"};
-  size_t nOptions = 3;
+  const char* compileOptions[] = {"-D__HIP", "-O3", "-w", "-munsafe-fp-atomics"};
```
[the remark this answers is not in the document]
Comment (gsitaram) [reply src/acc/libsmm_acc/libsmm_acc.cpp:144]: Added in src/acc/libsmm_acc/tune/tune_setup.py.

File: src/acc/libsmm_acc/calculate_norms.cpp [hunk 3]
```diff
@@ -0,0 +1,117 @@
+/*------------------------------------------------------------------------------------------------*/
```
[the remark this answers is not in the document]
Comment (gsitaram) [reply src/acc/libsmm_acc/calculate_norms.cpp]: Done this.

File: src/acc/libsmm_acc/calculate_norms.cpp [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+ * --- warp level primitives are used to reduce within a warp/wavefront, and
+ *     shared memory is used if more than one warp/wavefront is detected
+ */
+
+#if defined(__CUDA)
+#  include "../cuda/acc_cuda.h"
+#elif defined(__HIP)
+#  include "../hip/acc_hip.h"
+#endif
+#include "libsmm_acc_init.h"
```
[the remark this answers is not in the document]
Comment (gsitaram) [reply src/acc/libsmm_acc/calculate_norms.cpp]: I included libsmm_acc_init.h to be able to call `acc_get_gpu_warp_size()`. So, leaving this in there.

**Comment by gsitaram** [comment]
@alazzaro , please check if[ the previous commit ](https://github.com/cp2k/dbcsr/pull/647/commits/f2311f9b6aac090383f0c41f1af8978f0b6e90c3)reflects our discussion about avoiding the g2g path for all data types other than real_8.

**Comment by hfp** [comment]
I think Gina was right with the previous location of `c_calculate_norms` (under `src/acc/libsmm_acc`). That's the "library" also hosting transpose and multiply (in fact it's also CUDA/HIP-only). Another sign of `calculate_norms.cpp` belonging to LIBSMM_ACC is the latter only contains infrastructure-code like mem, stream, error, etc., and never contained calculation routines. Further, `calculate_norms.cpp` depends on `libsmm_acc_init.h` (`acc_get_gpu_warp_size`), which is now awkward after moving it to cuda_hip.

**Comment by alazzaro** [comment]
thanks @gsitaram !

**Pull request merged**